k8s/src/main/scala/io/buoyant/k8s/ServiceNamer.scala

```diff
@@ -23,17 +23,107 @@ class ServiceNamer(
+        // the address changes, *or* when the value of the address changes.
+        val unstable = service(nsName.toLowerCase, serviceName.toLowerCase, None)
+          .map { _.lookup(portName.toLowerCase) }
+        // stabilize the activity by converting it into an
+        // `Activity[Option[Var[Address]]]`, where the outer `Activity` will
+        // update if the `Option` changes, and the inner `Var` will update on
+        // changes to the value of the `Address`.
+        stabilize(unstable)
+          // convert the contents of the stable activity to a `NameTree`.
+          .map(toNameTree(path, _))
+      case (id@Path.Utf8(nsName, portName, serviceName, labelValue), Some(label)) =>
+        val labelSelector = Some(s"$label=$labelValue")
+        // as above, create an unstable activity, stabilize it, and then
+        // convert to a `NameTree`.
+        val unstable = service(
+          nsName.toLowerCase,
+          serviceName.toLowerCase,
+          labelSelector
+        ).map { _.lookup(portName.toLowerCase) }
+        stabilize(unstable).map(toNameTree(path, _))
+      case _ =>
+        Activity.value(NameTree.Neg)
+    }
+
+  private[this] def toNameTree(
+    path: Path,
+    svcAddress: Option[Var[Address]]
+  ): NameTree[Name.Bound] = svcAddress match {
+    case Some(address) =>
+      val residual = path.drop(variablePrefixLength)
+      val id = path.take(variablePrefixLength)
+      val bound = address.map(Addr.Bound(_))
+      NameTree.Leaf(Name.Bound(bound, idPrefix ++ id, residual))
+    case None =>
+      NameTree.Neg
+  }
+
+}
+private[this] object ServiceNamer {
+
+  val PrefixLen = 3
+
+  def unpackService(service: v1.Service): (Map[String, Address], Map[Int, String]) = {
+    val ports = mutable.Map.empty[String, Address]
+    val portMap = mutable.Map.empty[Int, String]
+
+    for {
+      meta <- service.metadata.toSeq
+      name <- meta.name.toSeq
+      status <- service.status.toSeq
+      lb <- status.loadBalancer.toSeq
+      spec <- service.spec.toSeq
+      v1.ServicePort(port, targetPort, name) <- spec.ports
+    } {
+      for {
+        ingress <- lb.ingress.toSeq.flatten
+        hostname <- ingress.hostname.orElse(ingress.ip)
+      } ports += name -> Address(new InetSocketAddress(hostname, port))
+
+      portMap += (targetPort match {
+        case Some(target) => port -> target
+        case None => port -> port.toString
+      })
+    }
+    (ports.toMap, portMap.toMap)
+  }
 
   /**
    * Internal representation of a Kubernetes service as a map of port names
    * to `Address`es and a map of port numbers to port names.
    * @param ports a map of `String`s representing port names to `Address`es.
    * @param portMappings a map of port numbers to port names.
    */
-  private[this] case class Svc(
+  case class Svc(
+    nsName: String,
+    serviceName: String,
     ports: Map[String, Address],
     portMappings: Map[Int, String]
-  ) {
+  ) extends PortMapLogging {
```

Comment:
would prefer to have the PortMapLogging as a private member rather than a parent.